// babel-plugin-transform-modules-ui5/dist/classes/helpers/classes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.convertClassToUI5Extend = convertClassToUI5Extend;
exports.getClassInfo = getClassInfo;
var _core = require("@babel/core");
var _path = _interopRequireDefault(require("path"));
var _objectAssignDefined = _interopRequireDefault(require("object-assign-defined"));
var th = _interopRequireWildcard(require("../../utils/templates"));
var ast = _interopRequireWildcard(require("../../utils/ast"));
var _jsdoc = require("./jsdoc");
var _decorators = require("./decorators");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
/**
 * Converts an ES6 class to a UI5 extend.
 * Any static methods or properties will be moved outside the class body.
 * The path will be updated with the new AST.
 */
function convertClassToUI5Extend(path, node, classInfo, extraStaticProps, opts) {
  if (!(_core.types.isClassDeclaration(node) || _core.types.isClassExpression(node))) {
    return node;
  }
  const staticMembers = [];
  const classNameIdentifier = node.id;
  const className = classNameIdentifier.name;
  const superClass = node.superClass; // Identifier node.
  const superClassName = superClass.name;
  const isController = className.includes("Controller") || !!classInfo.controller;
  const moveControllerConstructorToOnInit = isController && !!opts.moveControllerConstructorToOnInit;
  const moveControllerPropsToOnInit = isController && (!!opts.moveControllerPropsToOnInit || !!opts.moveControllerConstructorToOnInit);
  const moveStaticStaticPropsToExtend = isController && !!opts.addControllerStaticPropsToExtend;
  const alwaysMoveInstanceProps = !opts.onlyMoveClassPropsUsingThis;
  const extendProps = [];
  const boundProps = [];
  const CONSTRUCTOR = "constructor";
  const propsByName = {};
  let constructor;
  let constructorComments;
  const staticPropsToAdd = moveStaticStaticPropsToExtend ? Object.keys(extraStaticProps) : ["metadata", "renderer", "overrides"];
  for (const propName of staticPropsToAdd) {
    if (extraStaticProps[propName]) {
      extendProps.push(_core.types.objectProperty(_core.types.identifier(propName), extraStaticProps[propName]));
    }
  }
  for (const member of node.body.body) {
    const memberName = member.key.name;
    if (_core.types.isClassMethod(member)) {
      var _member$params;
      const isConstructor = member.kind === "constructor";
      const membersToAssign = [];
      const params = isConstructor ? (_member$params = member.params) === null || _member$params === void 0 ? void 0 : _member$params.map(param => {
        // handling of parameter properties for constructors (TypeScript):
        // https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties
        //   -> extracting the real parameters and store the members to assign
        if (param.type === "TSParameterProperty") {
          membersToAssign.push(param.parameter);
          return param.parameter;
        }
        return param;
      }) : member.params;
      const func = _core.types.functionExpression(member.key, params, member.body, member.generator, member.async);
      if (isConstructor && membersToAssign.length > 0) {
        // handling of parameter properties for constructors (TypeScript):
        //   -> assigning parameter properties as members to the instance
        const newMembers = membersToAssign.map(member => buildMemberAssignmentStatement(_core.types.thisExpression(), {
          key: member,
          computed: false,
          value: member
        }));
        const superIndex = member.body.body.findIndex(node => ast.isSuperCallExpression(node.expression) || ast.isSuperPrototypeCallOf(node.expression, superClassName, "constructor"));
        member.body.body.splice(superIndex === -1 ? member.body.body.length : superIndex + 1, 0, ...newMembers);
      }
      if (member.static) {
        staticMembers.push(buildMemberAssignmentStatement(classNameIdentifier, _objectSpread(_objectSpread({}, member), {}, {
          value: func
        })));
      } else {
        propsByName[memberName] = func;
        if (member.kind === "get" || member.kind === "set") {
          extendProps.push(_core.types.objectMethod(member.kind, member.key, member.params, member.body, member.computed));
        } else {
          // method
          if (memberName === CONSTRUCTOR) {
            constructorComments = member.leadingComments;
            constructor = func;
            if (moveControllerPropsToOnInit) {
              continue; // don't push to props yet
            }
          }

          func.id = path.scope.generateUidIdentifier(func.id.name); // Give the function a unique name
          extendProps.push(buildObjectProperty(_objectSpread(_objectSpread({}, member), {}, {
            value: func
          })));
        }
      }
    } else if (_core.types.isClassProperty(member)) {
      if (!member.value) continue; // un-initialized static class prop (typescript)

      // Special handling for TypeScript limitation where metadata, renderer and overrides must be properties.
      if (["metadata", "renderer", "overrides"].includes(memberName)) {
        if (opts.overridesToOverride && member.key.name === "overrides") {
          member.key.name = "override";
        }
        extendProps.unshift(buildObjectProperty(member));
      } else if (member.static) {
        if (moveStaticStaticPropsToExtend) {
          extendProps.unshift(buildObjectProperty(member));
        } else {
          staticMembers.push(buildMemberAssignmentStatement(classNameIdentifier, member));
        }
      } else {
        propsByName[memberName] = member.value;
        if (memberName === "constructor") {
          constructorComments = member.leadingComments;
          constructor = member.value;
          if (moveControllerPropsToOnInit) {
            continue; // don't push to props yet
          }
        }

        if (alwaysMoveInstanceProps || _core.types.isArrowFunctionExpression(member.value) || ast.isThisExpressionUsed(member.value)) {
          boundProps.push(member);
        } else {
          extendProps.push(buildObjectProperty(member));
        }
      }
    }
  }

  // Arrow function properties need to get moved to the constructor so that
  // they're bound properly to the class instance, to align with the spec.
  // For controllers, use onInit rather than constructor, since controller constructors don't work.
  // Also move the constructor's statements to the onInit.

  const bindToConstructor = !moveControllerPropsToOnInit;
  const bindToMethodName = moveControllerPropsToOnInit ? "onInit" : "constructor";

  // avoid getting a prop named constructor as it may return {}'s
  let bindMethod = moveControllerPropsToOnInit ? propsByName[bindToMethodName] : constructor;
  const constructorJsdoc = (0, _jsdoc.getTags)(constructorComments);
  const keepConstructor = !moveControllerConstructorToOnInit || classInfo.keepConstructor || constructorJsdoc.keep;

  // See if we need either constructor or onInit
  const needsBindingMethod = boundProps.length || moveControllerPropsToOnInit && constructor && !keepConstructor;

  // See if we need to create a new 'constructor' or 'onInit' method, depending which one we'll bind to.
  if (needsBindingMethod && !bindMethod) {
    const bindToId = _core.types.identifier(bindToMethodName);
    const bindMethodDeclaration = bindToConstructor ? th.buildInheritingConstructor({
      SUPER: _core.types.identifier(superClassName)
    }) : th.buildInheritingFunction({
      NAME: bindToId,
      SUPER: _core.types.identifier(superClassName)
    });
    bindMethod = ast.convertFunctionDeclarationToExpression(bindMethodDeclaration);
    extendProps.unshift(_core.types.objectProperty(bindToId, bindMethod));
  }
  if (constructor && moveControllerPropsToOnInit) {
    if (keepConstructor) {
      extendProps.unshift(_core.types.objectProperty(_core.types.identifier(CONSTRUCTOR), constructor));
    } else {
      // Copy all except the super call from the constructor to the bindMethod (i.e. onInit)
      bindMethod.body.body.unshift(...constructor.body.body.filter(node => !(ast.isSuperCallExpression(node.expression) || ast.isSuperPrototypeCallOf(node.expression, superClassName, "constructor"))));
    }
  }
  if (boundProps.length) {
    // We need to inject the bound props into the bind method (constructor or onInit),
    // but not until after the super call (if applicable)

    const mappedProps = boundProps.map(member => buildThisMemberAssignmentStatement(member));
    const superIndex = bindMethod.body.body.findIndex(node => ast.isSuperCallExpression(node.expression) || ast.isSuperPrototypeCallOf(node.expression, superClassName, bindToMethodName));
    if (superIndex === -1) {
      // If there's no super, just add the bound props at the start
      bindMethod.body.body.unshift(...mappedProps);
    } else {
      const upToSuper = bindMethod.body.body.slice(0, superIndex + 1);
      const afterSuper = bindMethod.body.body.slice(superIndex + 1);
      bindMethod.body.body = [...upToSuper, ...mappedProps, ...afterSuper];
    }
  }
  const extendAssign = th.buildExtendAssign({
    NAME: classNameIdentifier,
    SUPER: superClass,
    // Needs Identifier node
    FQN: _core.types.stringLiteral(getFullyQualifiedName(classInfo)),
    OBJECT: _core.types.objectExpression(extendProps)
  });
  return [extendAssign, ...staticMembers];
}
function getFullyQualifiedName(classInfo) {
  if (classInfo.alias) return classInfo.alias;
  if (classInfo.name) return classInfo.name;
  const namespace = classInfo.namespace || classInfo.fileNamespace;
  const separator = namespace ? "." : "";
  return `${namespace}${separator}${classInfo.localName}`;
}
function getClassInfo(path, node, parent, pluginOpts) {
  const defaults = {
    localName: node.id.name,
    superClassName: node.superClass && node.superClass.name,
    fileNamespace: getFileBaseNamespace(path, pluginOpts) || ""
  };
  const decoratorInfo = (0, _decorators.getDecoratorClassInfo)(node);
  const jsDocInfo = (0, _jsdoc.getJsDocClassInfo)(node, parent);

  // like Object.assign, but ignoring undefined values.
  return (0, _objectAssignDefined.default)(defaults, decoratorInfo, jsDocInfo);
}

/**
 * Reads the namespace from the file path (but not the name).
 */
function getFileBaseNamespace(path, pluginOpts) {
  const opts = path.hub.file.opts;
  const filename = _path.default.resolve(opts.filename);
  const sourceRoot = opts.sourceRoot ? _path.default.resolve(process.cwd(), opts.sourceRoot) : process.cwd();
  if (filename.startsWith(sourceRoot)) {
    const filenameRelative = _path.default.relative(sourceRoot, filename);
    const {
      dir
    } = _path.default.parse(filenameRelative);
    const namespaceParts = dir.split(_path.default.sep);
    if (pluginOpts.namespacePrefix) {
      namespaceParts.unshift(pluginOpts.namespacePrefix);
    }
    return namespaceParts.join(".");
  } else {
    return undefined;
  }
}
const buildObjectProperty = member => {
  const newObjectProperty = _core.types.objectProperty(member.key, member.value, member.computed);
  newObjectProperty.leadingComments = member.leadingComments;
  return newObjectProperty;
};
const buildMemberAssignmentStatement = (objectIdentifier, member) => {
  const newMember = _core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(objectIdentifier, member.key, member.computed), member.value));
  newMember.leadingComments = member.leadingComments;
  return newMember;
};
const buildThisMemberAssignmentStatement = buildMemberAssignmentStatement.bind(null, _core.types.thisExpression());
